File: fugaku_in_practice/hands_on/sample_code/01_processor-core/01_polynomial/cpp/mykernel.cpp
// Copyright 2024 Research Organization for Information Science and Technology
#include "mykernel.h"

void poly1 (double * _restrict_cpp x, double * _restrict_cpp c0, 
            double * _restrict_cpp c1, const int ndim)
{
    for( int i=0; i<ndim; ++i ) {
      x[i] = c0[i]+x[i]*c1[i] ;
    }
}

void poly4 (double * _restrict_cpp x, double * _restrict_cpp c0, 
            double * _restrict_cpp c1, const int ndim)
{
    for( int i=0; i<ndim; ++i ) {
      x[i] = c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              )))) ;
    }
}

void poly8 (double * _restrict_cpp x, double * _restrict_cpp c0, 
            double * _restrict_cpp c1, const int ndim)
{
    for( int i=0; i<ndim; i++ ) {
      x[i] = c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              )))))))) ;
    }
}

void poly16 (double * _restrict_cpp x, double * _restrict_cpp c0, 
            double * _restrict_cpp c1, const int ndim)
{
    for( int i=0; i<ndim; ++i ) {
      x[i] = c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              +x[i]*(c1[i] +x[i]*(c0[i] +x[i]*(c1[i] +x[i]*(c0[i]
              )))))))))))))))) ;
    }
}

File: fugaku_in_practice/hands_on/sample_code/01_processor-core/01_polynomial/fortran_DoConcurrent/mykernel.f90
! Copyright 2024 Research Organization for Information Science and Technology
module mykernel
  use mytype, only: DP
  implicit none
contains
  subroutine poly1(ndim, x, c1, c2)
     integer, intent(in) :: ndim
     real(kind=DP), intent(inout) :: x(ndim)
     real(kind=DP), intent(in) :: c1(ndim)
     real(kind=DP), intent(in) :: c2(ndim)
     
     integer :: i
     do concurrent (i=1:ndim)
        x(i) = c1(i)+x(i)*c2(i)
     end do
  end subroutine poly1

  subroutine poly4(ndim, x, c1, c2)
     integer, intent(in) :: ndim
     real(kind=DP), intent(inout) :: x(ndim)
     real(kind=DP), intent(in) :: c1(ndim)
     real(kind=DP), intent(in) :: c2(ndim)
     
     integer :: i
     do concurrent (i=1:ndim)
        x(i) = c1(i)                                               &
              +x(i)*(c2(i)+x(i)*(c1(i)+x(i)*(c2(i)+x(i)*c1(i)      &
              )))
     end do
  end subroutine poly4

  subroutine poly8(ndim, x, c1, c2)
     integer, intent(in) :: ndim
     real(kind=DP), intent(inout) :: x(ndim)
     real(kind=DP), intent(in) :: c1(ndim)
     real(kind=DP), intent(in) :: c2(ndim)
     
     integer :: i
     do concurrent (i=1:ndim)
        x(i) = c1(i)                                               &
              +x(i)*(c2(i)+x(i)*(c1(i)+x(i)*(c2(i)+x(i)*(c1(i)     &
              +x(i)*(c2(i)+x(i)*(c1(i)+x(i)*(c2(i)+x(i)*c1(i)      &
              )))))))
     end do
  end subroutine poly8

  subroutine poly16(ndim, x, c1, c2)
     integer, intent(in) :: ndim
     real(kind=DP), intent(inout) :: x(ndim)
     real(kind=DP), intent(in) :: c1(ndim)
     real(kind=DP), intent(in) :: c2(ndim)
     
     integer :: i
     do concurrent (i=1:ndim)
        x(i) = c1(i)                                               &
              +x(i)*(c2(i) +x(i)*(c1(i) +x(i)*(c2(i) +x(i)*( c1(i) &
              +x(i)*(c2(i) +x(i)*(c1(i) +x(i)*(c2(i) +x(i)*( c1(i) &
              +x(i)*(c2(i) +x(i)*(c1(i) +x(i)*(c2(i) +x(i)*( c1(i) &
              +x(i)*(c2(i) +x(i)*(c1(i) +x(i)*(c2(i) +x(i)*( c1(i) &
              ))))))))))))))))
     end do
  end subroutine poly16
end module mykernel
